### DISCUSSION TITLE
Not understanding how ".intend" internal action works.

### DISCUSSION THREAD

**Discussion by mindcrime** [jason-lang/jason#121] [Q&A] [answered]
Folks:

I'm reading Chapter 8 of the "Programming Multi-agent Systems in AgentSpeak with Jason" book, and have run into a sticking point. After reading about .desire, .intend, .drop_desire, .drop_intention, etc., I had intended (heh, heh) to write some sample code to experiment with these different methods.

Doing .desire worked OK, doing something like this:

agent2.asl
```
!started.

!achieve_stuff.

+!achieve_stuff : true <- .wait(1500); .print( "Well frack me" ); .print( "Whatever").

+!started <- .desire(achieve_stuff); .print( "Desire exists!" ).
```

I get 

```
[agent2] Desire exists!
[agent2] Well frack me
[agent2] Whatever
```

But if I try to test for an Intention instead, with .intend, using something like this:

agent1.asl
```
!started.
!started2.

!achieve_stuff.

+!achieve_stuff : true <- .wait(1500); .print( "Well frack me" ); !foo; .print( "Whatever").

+!foo : true <- .wait(6000); .print( "OK, DONE" ).

+!started <- .desire(achieve_stuff); .print( "Desire exists!" ).

+!started2 : .intend(foo, I ) <- .print( "Intention exists!\n", I ).
```

I get this:

```
[agent1] No fail event was generated for +!started2[source(self)]
[agent1] Found a goal for which there is no applicable plan: +!started2[source(self)]
[agent1] Desire exists!
[agent1] Well frack me
[agent1] OK, DONE
[agent1] Whatever
```

So it looks like my .intend is returning false, even though I expected it to return true.  Well , sort of. Truth be told, I don't exactly understand what the documentation here is telling me, so was hoping that maybe there was just a typo, but apparently not.

In the [docs](https://jason-lang.github.io/api/jason/stdlib/intend.html), it reads:

```
{".intend(go(1,3)): is true if a plan with triggering event +!go(1,3) appears in an intention of the agent
```

I don't quite know what it means to say "triggering event +!go(1,3) appears in an intention". I think intentions are selected plans, but the syntax +!foo in the body of a plan doesn't seem to be valid syntax. And I thought maybe just having a !foo in a plan, with a corresponding triggering event of +!foo might be enough, but clearly not.

Can somebody tell me what it is that I'm missing here?  It would be greatly appreciated.

**Comment by jomifred** [comment, marked as answer]
wow, very nice example!

if I’ve got it right, when the test `.intend…` is executed, the agent does not indent foo yet, it will  ~1500ms later.

this [doc](https://github.com/jason-lang/jason/blob/main/doc/tech/goal-states.pdf) may be of your interest.

**Reply by mindcrime** [reply]
OK, I was thinking it might be a timing issue. I did try rewriting it like this, so that the !foo happened earlier:

```
!started.
!started2.

!achieve_stuff.

+!achieve_stuff : true <- !foo; .print( "Whatever").

+!foo : true <- .wait(8000); .print( "OK, DONE" ).

+!started <- .desire(achieve_stuff); .print( "Desire exists!" ).

+!started2 : true <- .intend(foo); .print( "Intention exists!\n" ).
```

I also moved the .intend()  into the plan body and took out the variable to unify with the Intention, but I still get:

```
[agent1] Desire exists!
[agent1] No failure event was generated for +!started2[code(.intend(foo)),code_line(12),code_src("file:src/agt/agent1.asl"),error(ia_failed),error_msg(""),source(self)]
intention 2: 
    +!started2[source(self)] <- ... .intend(foo); .print("Intention exists!
") / {}

[agent1] OK, DONE
[agent1] Whatever
```

> this [doc](https://github.com/jason-lang/jason/blob/main/doc/tech/goal-states.pdf) may be of your interest.

OK, I'll give that a read.  Thanks for the pointer!

**Reply by mindcrime** [reply]
OK, wait. I think I get it now. Definitely a timing issue. I realized the .intend() could still be firing before the foo plan (intention) was selected, so I added some time *before* the .intend(foo) bit. Like this:

```
!started.
!started2.

!achieve_stuff.

+!achieve_stuff : true <- !foo; .print( "Whatever").

+!foo : true <- .wait(8000); .print( "OK, DONE" ).

+!started <- .desire(achieve_stuff); .print( "Desire exists!" ).

+!started2 : true <- .wait(500); .intend(foo); .print( "Intention exists!\n" ).
```

and now I get this:

```
[agent1] Desire exists!
[agent1] Intention exists!
```

I think what I was missing earlier was the realization of just how critical the timing part was, and the point that without some specific wait()'ing in there my one plan could still run before the other one and cause the test to miss the actual Intention. So intend() really is saying "is the corresponding plan/intention selected *right now*".
